File: Problems/No.1871_Jump_Game_VII/jump_game_VII.py
class Solution(object):
    def canReach(self, s, minJump, maxJump):
        """
        :type s: str
        :type minJump: int
        :type maxJump: int
        :rtype: bool
        """

        # parameters
        s_length = len(s)
        index = []

        # start computing
        if (s[-1] == '1'):
            return False
        MAX_idx = 0
        index.append(0)
        while index:
            idx = index.pop(0)
            for i in range(max(idx+minJump, MAX_idx+1), min(s_length, idx+maxJump+1)):
                if (s[i] == '0'):
                    if (i == s_length-1):
                        return True
                    index.append(i)
            MAX_idx = idx + maxJump
        
        return False

File: Problems/No.1871_Jump_Game_VII/jump_game_VII.cpp
class Solution {
public:
    bool canReach(string s, int minJump, int maxJump) {
        
        // parameters
        int s_length = s.length();
        std::queue<int> index;

        // start computing
        int MAX_idx = 0;
        index.push(0);
        while (!index.empty()) {

            int idx = index.front();
            index.pop();
            
            for (int i=std::max(idx+minJump, MAX_idx+1);
                     i<std::min(s_length, idx+maxJump+1); i++)
                if (s[i] == '0') {

                    if (i == s_length-1) return true;
                    index.push(i);

                }
            
            MAX_idx = idx + maxJump;

        }

        return false;

    }
};
